platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/service/WritableRosterStore.java

```diff
@@ -0,0 +1,181 @@
+        requireNonNull(roster);
+        RosterValidator.validate(roster);
+
+        // update the roster state
+        final RosterState previousRosterState = rosterStateOrThrow();
+        final List<RoundRosterPair> roundRosterPairs = new LinkedList<>(previousRosterState.roundRosterPairs());
+        if (!roundRosterPairs.isEmpty()) {
+            final RoundRosterPair activeRosterPair = roundRosterPairs.getFirst();
+            if (round < 0 || round <= activeRosterPair.roundNumber()) {
+                throw new IllegalArgumentException(
+                        "incoming round number must be greater than the round number of the current active roster.");
+            }
+        }
+        final Bytes activeRosterHash = RosterUtils.hash(roster).getBytes();
+        roundRosterPairs.addFirst(new RoundRosterPair(round, activeRosterHash));
+
+        if (roundRosterPairs.size() > MAXIMUM_ROSTER_HISTORY_SIZE) {
+            final RoundRosterPair lastRemovedRoster = roundRosterPairs.removeLast();
+            removeRoster(lastRemovedRoster.activeRosterHash());
+
+            // At this phase of the implementation, the roster state has a fixed size limit for active rosters.
+            // Future implementations (e.g. DAB) can modify this.
+            if (roundRosterPairs.size() > MAXIMUM_ROSTER_HISTORY_SIZE) {
+                // additional safety check to ensure that the roster state does not contain more than set limit.
+                throw new IllegalStateException(
+                        "Active rosters in the Roster state cannot be more than  " + MAXIMUM_ROSTER_HISTORY_SIZE);
+            }
+        }
+
+        final Builder newRosterState = previousRosterState
+                .copyBuilder()
+                .candidateRosterHash(Bytes.EMPTY)
+                .roundRosterPairs(roundRosterPairs);
+        // since a new active roster is being set, the existing candidate roster is no longer valid
+        // so we remove it if it meets removal criteria.
+        removeRoster(previousRosterState.candidateRosterHash());
+        storeRoster(roster, activeRosterHash, newRosterState);
+    }
+
+    /**
+     * Returns the roster state or throws an exception if the state is null.
+     * @return the roster state
+     * @throws NullPointerException if the roster state is null
+     */
+    @NonNull
+    private RosterState rosterStateOrThrow() {
+        return requireNonNull(rosterState.get());
+    }
+
+    /**
+     * Stores the roster in the roster map and updates the roster store with the provided builder.
+     *
+     * @param rosterToStore              the roster to store
+     * @param rosterHash                   the hash of the roster
+     * @param rosterStateBuilder       the roster state builder
+     */
+    private void storeRoster(
+            @NonNull final Roster rosterToStore,
+            @NonNull final Bytes rosterHash,
+            @NonNull final Builder rosterStateBuilder) {
```

Comment:
This is a dangerous method. If one forgets to use the `copyBuilder()` on a previous roster state, then they can accidentally overwrite a candidate roster hash or the roster history depending on how they configured the builder. I'd suggest to eliminate this method completely to avoid any bugs in the future. It's very short and doesn't contain any logic. It can be replicated inline easily, and be safe this way because the updates to the state will be explicit.

Reply:
Removed this method and inlined in usages.